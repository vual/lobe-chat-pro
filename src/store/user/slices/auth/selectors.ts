import { t } from 'i18next';

import { enableAuth, enableClerk, enableNextAuth } from '@/const/auth';
import { BRANDING_NAME } from '@/const/branding';
import { isDesktop } from '@/const/version';
import { UserStore } from '@/store/user';
import { LobeUser } from '@/types/user';

const DEFAULT_USERNAME = BRANDING_NAME;

const nickName = (s: UserStore) => {
  const defaultNickName = s.user?.fullName || s.user?.username;
  if (!enableAuth) {
    if (isDesktop) {
      return defaultNickName;
    }
    return t('userPanel.defaultNickname', { ns: 'common' });
  }

  if (s.isSignedIn) return defaultNickName;

  return t('userPanel.anonymousNickName', { ns: 'common' });
};

const username = (s: UserStore) => {
  if (!enableAuth) {
    if (isDesktop) {
      return s.user?.username;
    }
    return DEFAULT_USERNAME;
  }

  if (s.isSignedIn) return s.user?.username;

  return 'anonymous';
};

export const userProfileSelectors = {
  displayUserName: (s: UserStore): string => username(s) || s.user?.email || '',
  email: (s: UserStore): string => s.user?.email || '',
  fullName: (s: UserStore): string => s.user?.fullName || '',
  nickName,
  userAvatar: (s: UserStore): string => s.user?.avatar || '',
  userId: (s: UserStore) => s.user?.id,
  userProfile: (s: UserStore): LobeUser | null | undefined => s.user,
  username,
};

/**
 * 使用此方法可以兼容不需要登录鉴权的情况
 */
const isLogin = (s: UserStore) => {
  // 如果没有开启鉴权，说明不需要登录，默认是登录态
  if (!enableAuth) return true;

  return s.isSignedIn;
};

export const authSelectors = {
  isLoaded: (s: UserStore) => s.isLoaded,
  isLogin,
  isLoginWithAuth: (s: UserStore) => s.isSignedIn,
  isLoginWithClerk: (s: UserStore): boolean => (s.isSignedIn && enableClerk) || false,
  isLoginWithNextAuth: (s: UserStore): boolean => (s.isSignedIn && !!enableNextAuth) || false,
};
