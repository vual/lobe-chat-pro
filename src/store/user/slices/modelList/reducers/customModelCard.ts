import { produce } from 'immer';

import { ChatModelCard } from '@/types/llm';

export interface AddCustomModelCard {
  modelCard: ChatModelCard;
  type: 'add';
}

export interface DeleteCustomModelCard {
  id: string;
  type: 'delete';
}

export interface UpdateCustomModelCard {
  id: string;
  type: 'update';
  value: Partial<ChatModelCard>;
}

export type CustomModelCardDispatch =
  | AddCustomModelCard
  | DeleteCustomModelCard
  | UpdateCustomModelCard;

export const customModelCardsReducer = (
  state: ChatModelCard[] | undefined,
  payload: CustomModelCardDispatch,
): ChatModelCard[] => {
  switch (payload.type) {
    case 'add': {
      return produce(state || [], (draftState) => {
        const { id } = payload.modelCard;
        if (!id) return;
        if (draftState.some((card) => card.id === id)) return;

        draftState.push(payload.modelCard);
      });
    }

    case 'delete': {
      return produce(state || [], (draftState) => {
        const index = draftState.findIndex((card) => card.id === payload.id);
        if (index !== -1) {
          draftState.splice(index, 1);
        }
      });
    }

    case 'update': {
      return produce(state || [], (draftState) => {
        const index = draftState.findIndex((card) => card.id === payload.id);
        if (index !== -1) {
          const card = draftState[index];
          Object.assign(card, payload.value);
        }
      });
    }

    default: {
      throw new Error('Unhandled action type in customModelCardsReducer');
    }
  }
};
