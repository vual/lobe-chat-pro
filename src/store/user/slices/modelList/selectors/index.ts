export { keyVaultsConfigSelectors } from './keyVaults';
export { modelConfigSelectors } from './modelConfig';
export { modelProviderSelectors } from './modelProvider';
