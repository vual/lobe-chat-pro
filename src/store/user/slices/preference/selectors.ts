import { DEFAULT_PREFERENCE } from '@/const/user';
import { UserStore } from '@/store/user';

const useCmdEnterToSend = (s: UserStore): boolean => s.preference.useCmdEnterToSend || false;
const topicDisplayMode = (s: UserStore) =>
  s.preference.topicDisplayMode || DEFAULT_PREFERENCE.topicDisplayMode;

const userAllowTrace = (s: UserStore) => s.preference.telemetry;

const hideSyncAlert = (s: UserStore) => s.preference.hideSyncAlert;

const hideSettingsMoveGuide = (s: UserStore) => s.preference.guide?.moveSettingsToAvatar;

const showUploadFileInKnowledgeBaseTip = (s: UserStore) =>
  s.preference.guide?.uploadFileInKnowledgeBase;

const shouldTriggerFileInKnowledgeBaseTip = (s: UserStore) =>
  !(typeof s.preference.guide?.moveSettingsToAvatar === 'boolean');

const isPreferenceInit = (s: UserStore) => s.isUserStateInit;

export const preferenceSelectors = {
  hideSettingsMoveGuide,
  hideSyncAlert,
  isPreferenceInit,
  shouldTriggerFileInKnowledgeBaseTip,
  showUploadFileInKnowledgeBaseTip,
  topicDisplayMode,
  useCmdEnterToSend,
  userAllowTrace,
};
