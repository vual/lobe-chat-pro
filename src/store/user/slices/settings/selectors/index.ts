export { userGeneralSettingsSelectors } from './general';
export { settingsSelectors } from './settings';
export { systemAgentSelectors } from './systemAgent';
