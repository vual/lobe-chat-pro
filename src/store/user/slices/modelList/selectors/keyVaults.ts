import { UserStore } from '@/store/user';
import {
  AWSBedrockKeyVault,
  AzureOpenAIKeyVault,
  GlobalLLMProviderKey,
  OpenAICompatibleKeyVault,
  UserKeyVaults,
} from '@/types/user/settings';

import { currentSettings } from '../../settings/selectors/settings';

export const keyVaultsSettings = (s: UserStore): UserKeyVaults =>
  currentSettings(s).keyVaults || {};

const openAIConfig = (s: UserStore) => keyVaultsSettings(s).openai || {};
const bedrockConfig = (s: UserStore) => keyVaultsSettings(s).bedrock || {};
const ollamaConfig = (s: UserStore) => keyVaultsSettings(s).ollama || {};
const azureConfig = (s: UserStore) => keyVaultsSettings(s).azure || {};
const cloudflareConfig = (s: UserStore) => keyVaultsSettings(s).cloudflare || {};
const getVaultByProvider = (provider: GlobalLLMProviderKey) => (s: UserStore) =>
  (keyVaultsSettings(s)[provider] || {}) as OpenAICompatibleKeyVault &
    AzureOpenAIKeyVault &
    AWSBedrockKeyVault;

const isProviderEndpointNotEmpty = (provider: string) => (s: UserStore) => {
  const vault = getVaultByProvider(provider as GlobalLLMProviderKey)(s);
  return !!vault?.baseURL || !!vault?.endpoint;
};

const isProviderApiKeyNotEmpty = (provider: string) => (s: UserStore) => {
  const vault = getVaultByProvider(provider as GlobalLLMProviderKey)(s);
  return !!vault?.apiKey || !!vault?.accessKeyId || !!vault?.secretAccessKey;
};

const password = (s: UserStore) => keyVaultsSettings(s).password || '';

export const keyVaultsConfigSelectors = {
  azureConfig,
  bedrockConfig,
  cloudflareConfig,
  getVaultByProvider,
  isProviderApiKeyNotEmpty,
  isProviderEndpointNotEmpty,
  keyVaultsSettings,
  ollamaConfig,
  openAIConfig,
  password,
};
