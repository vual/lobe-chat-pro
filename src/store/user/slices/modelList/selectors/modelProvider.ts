import { uniqBy } from 'lodash-es';

import { filterEnabledModels } from '@/config/modelProviders';
import { EnabledProviderWithModels } from '@/types/aiProvider';
import { ChatModelCard, ModelProviderCard } from '@/types/llm';
import { ServerModelProviderConfig } from '@/types/serverConfig';
import { GlobalLLMProviderKey } from '@/types/user/settings';

import { UserStore } from '../../../store';
import { currentSettings, getProviderConfigById } from '../../settings/selectors/settings';

/**
 * get the server side model cards
 */
const serverProviderModelCards =
  (provider: GlobalLLMProviderKey) =>
  (s: UserStore): ChatModelCard[] | undefined => {
    const config = s.serverLanguageModel?.[provider] as ServerModelProviderConfig | undefined;

    if (!config) return;

    return config.serverModelCards;
  };

const remoteProviderModelCards =
  (provider: GlobalLLMProviderKey) =>
  (s: UserStore): ChatModelCard[] | undefined => {
    const cards = currentSettings(s).languageModel?.[provider]?.remoteModelCards as
      | ChatModelCard[]
      | undefined;

    if (!cards) return;

    return cards;
  };

const isProviderEnabled = (provider: GlobalLLMProviderKey) => (s: UserStore) =>
  getProviderConfigById(provider)(s)?.enabled || false;

// Default Model Provider List

/**
 * define all the model list of providers
 */
const defaultModelProviderList = (s: UserStore): ModelProviderCard[] => s.defaultModelProviderList;

export const getDefaultModeProviderById = (provider: string) => (s: UserStore) =>
  defaultModelProviderList(s).find((s) => s.id === provider);

/**
 * get the default enabled models for a provider
 * it's a default enabled model list by Lobe Chat
 * e.g. openai is ['gpt-4o-mini','gpt-4o','gpt-4-turbo']
 */
const getDefaultEnabledModelsById = (provider: string) => (s: UserStore) => {
  const modelProvider = getDefaultModeProviderById(provider)(s);

  if (modelProvider) return filterEnabledModels(modelProvider);

  return undefined;
};

const getDefaultModelCardById = (id: string) => (s: UserStore) => {
  const list = defaultModelProviderList(s);

  return list.flatMap((i) => i.chatModels).find((m) => m.id === id);
};

// Model Provider List

const getModelCardsById =
  (provider: string) =>
  (s: UserStore): ChatModelCard[] => {
    const builtinCards = getDefaultModeProviderById(provider)(s)?.chatModels || [];

    const userCards = (getProviderConfigById(provider)(s)?.customModelCards || []).map((model) => ({
      ...model,
      isCustom: true,
    }));

    return uniqBy([...userCards, ...builtinCards], 'id');
  };

const getEnableModelsById = (provider: string) => (s: UserStore) => {
  if (!getProviderConfigById(provider)(s)?.enabledModels) return;

  return getProviderConfigById(provider)(s)?.enabledModels?.filter(Boolean);
};

const modelProviderList = (s: UserStore): ModelProviderCard[] => s.modelProviderList;

const modelProviderListForModelSelect = (s: UserStore): EnabledProviderWithModels[] =>
  modelProviderList(s)
    .filter((s) => s.enabled)
    .map((provider) => ({
      ...provider,
      children: provider.chatModels
        .filter((model) => model.enabled)
        .map((m) => ({
          abilities: {
            functionCall: m.functionCall,
            vision: m.vision,
          },
          contextWindowTokens: m.contextWindowTokens,
          displayName: m.displayName,
          id: m.id,
        })),
      source: 'builtin',
    }));

const getModelCardById = (id: string, provider?: GlobalLLMProviderKey) => (s: UserStore) => {
  const list = modelProviderList(s);

  return list
    .filter((i) => !provider || i.id === provider)
    .flatMap((i) => i.chatModels)
    .find((m) => m.id === id);
};

const isModelEnabledFunctionCall = (id: string) => (s: UserStore) =>
  getModelCardById(id)(s)?.functionCall || false;

// vision model white list, these models will change the content from string to array
// refs: https://github.com/lobehub/lobe-chat/issues/790
const isModelEnabledVision = (id: string) => (s: UserStore) =>
  getModelCardById(id)(s)?.vision || id.includes('vision');

const isModelEnabledReasoning = (id: string) => (s: UserStore) =>
  getModelCardById(id)(s)?.reasoning || false;

const isModelEnabledFiles = (id: string) => (s: UserStore) => getModelCardById(id)(s)?.files;

const isModelEnabledUpload = (id: string) => (s: UserStore) =>
  isModelEnabledVision(id)(s) || isModelEnabledFiles(id)(s);

const isModelHasMaxToken = (id: string) => (s: UserStore) =>
  typeof getModelCardById(id)(s)?.contextWindowTokens !== 'undefined';

const modelMaxToken = (id: string) => (s: UserStore) =>
  getModelCardById(id)(s)?.contextWindowTokens || 0;

export const modelProviderSelectors = {
  defaultModelProviderList,
  getDefaultEnabledModelsById,
  getDefaultModelCardById,

  getEnableModelsById,
  getModelCardById,

  getModelCardsById,
  isModelEnabledFiles,
  isModelEnabledFunctionCall,
  isModelEnabledReasoning,
  isModelEnabledUpload,
  isModelEnabledVision,
  isModelHasMaxToken,

  isProviderEnabled,

  modelMaxToken,
  modelProviderList,

  modelProviderListForModelSelect,

  remoteProviderModelCards,
  serverProviderModelCards,
};
