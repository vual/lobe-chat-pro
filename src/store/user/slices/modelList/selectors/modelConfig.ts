import { isProviderDisableBrowserRequest } from '@/config/modelProviders';
import { isDesktop } from '@/const/version';
import { UserStore } from '@/store/user';
import { GlobalLLMProviderKey } from '@/types/user/settings';

import { currentLLMSettings, getProviderConfigById } from '../../settings/selectors/settings';
import { keyVaultsConfigSelectors } from './keyVaults';

const isProviderEnabled = (provider: GlobalLLMProviderKey) => (s: UserStore) =>
  getProviderConfigById(provider)(s)?.enabled || false;

const providerWhitelist = new Set(['ollama']);
/**
 * @description The conditions to enable client fetch
 * 1. If no baseUrl and apikey input, force on Server.
 * 2. If only contains baseUrl, force on Client
 * 3. Follow the user settings.
 * 4. On Server, by default.
 */
const isProviderFetchOnClient = (provider: GlobalLLMProviderKey | string) => (s: UserStore) => {
  const config = getProviderConfigById(provider)(s);

  // if is desktop, force on Server.
  if (isDesktop) return false;

  // If the provider already disable browser request in model config, force on Server.
  if (isProviderDisableBrowserRequest(provider)) return false;

  // If the provider in the whitelist, follow the user settings
  if (providerWhitelist.has(provider) && typeof config?.fetchOnClient !== 'undefined')
    return config?.fetchOnClient;

  // 1. If no baseUrl and apikey input, force on Server.
  const isProviderEndpointNotEmpty =
    keyVaultsConfigSelectors.isProviderEndpointNotEmpty(provider)(s);
  const isProviderApiKeyNotEmpty = keyVaultsConfigSelectors.isProviderApiKeyNotEmpty(provider)(s);
  if (!isProviderEndpointNotEmpty && !isProviderApiKeyNotEmpty) return false;

  // 2. If only contains baseUrl, force on Client
  if (isProviderEndpointNotEmpty && !isProviderApiKeyNotEmpty) return true;

  // 3. Follow the user settings.
  if (typeof config?.fetchOnClient !== 'undefined') return config?.fetchOnClient;

  // 4. On Server, by default.
  return false;
};

const getCustomModelCard =
  ({ id, provider }: { id?: string; provider?: string }) =>
  (s: UserStore) => {
    if (!provider) return;

    const config = getProviderConfigById(provider)(s);

    return config?.customModelCards?.find((m) => m.id === id);
  };

const currentEditingCustomModelCard = (s: UserStore) => {
  if (!s.editingCustomCardModel) return;
  const { id, provider } = s.editingCustomCardModel;

  return getCustomModelCard({ id, provider })(s);
};

const isAutoFetchModelsEnabled =
  (provider: GlobalLLMProviderKey) =>
  (s: UserStore): boolean => {
    return getProviderConfigById(provider)(s)?.autoFetchModelLists || false;
  };

const openAIConfig = (s: UserStore) => currentLLMSettings(s).openai;
const bedrockConfig = (s: UserStore) => currentLLMSettings(s).bedrock;
const ollamaConfig = (s: UserStore) => currentLLMSettings(s).ollama;
const azureConfig = (s: UserStore) => currentLLMSettings(s).azure;
const cloudflareConfig = (s: UserStore) => currentLLMSettings(s).cloudflare;

const isAzureEnabled = (s: UserStore) => currentLLMSettings(s).azure.enabled;

export const modelConfigSelectors = {
  azureConfig,
  bedrockConfig,
  cloudflareConfig,

  currentEditingCustomModelCard,
  getCustomModelCard,

  isAutoFetchModelsEnabled,
  isAzureEnabled,
  isProviderEnabled,
  isProviderFetchOnClient,

  ollamaConfig,
  openAIConfig,
};
