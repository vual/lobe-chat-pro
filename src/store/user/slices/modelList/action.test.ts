import { act, renderHook, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { modelsService } from '@/services/models';
import { userService } from '@/services/user';
import { useUserStore } from '@/store/user';
import { ProviderConfig } from '@/types/user/settings';

import { settingsSelectors } from '../settings/selectors';
import { CustomModelCardDispatch } from './reducers/customModelCard';
import { modelProviderSelectors } from './selectors';

// Mock userService
vi.mock('@/services/user', () => ({
  userService: {
    updateUserSettings: vi.fn(),
    resetUserSettings: vi.fn(),
  },
}));

vi.mock('zustand/traditional');

describe('LLMSettingsSliceAction', () => {
  describe('setModelProviderConfig', () => {
    it('should set OpenAI configuration', async () => {
      const { result } = renderHook(() => useUserStore());
      const openAIConfig: Partial<ProviderConfig> = { fetchOnClient: true };

      // Perform the action
      await act(async () => {
        await result.current.setModelProviderConfig('openai', openAIConfig);
      });

      // Assert that updateUserSettings was called with the correct OpenAI configuration
      expect(userService.updateUserSettings).toHaveBeenCalledWith(
        { languageModel: { openai: openAIConfig } },
        expect.any(AbortSignal),
      );
    });
  });

  describe('dispatchCustomModelCards', () => {
    it('should return early when prevState does not exist', async () => {
      const { result } = renderHook(() => useUserStore());
      const provider = 'openai';
      const payload: CustomModelCardDispatch = { type: 'add', modelCard: { id: 'test-id' } };

      // Mock the selector to return undefined
      vi.spyOn(settingsSelectors, 'providerConfig').mockReturnValueOnce(() => undefined);
      vi.spyOn(result.current, 'setModelProviderConfig');

      await act(async () => {
        await result.current.dispatchCustomModelCards(provider, payload);
      });

      // Assert that setModelProviderConfig was not called
      expect(result.current.setModelProviderConfig).not.toHaveBeenCalled();
    });
  });

  describe('refreshDefaultModelProviderList', () => {
    it('default', async () => {
      const { result } = renderHook(() => useUserStore());

      act(() => {
        useUserStore.setState({
          serverLanguageModel: {
            azure: { serverModelCards: [{ id: 'abc', deploymentName: 'abc' }] },
          },
        });
      });

      act(() => {
        result.current.refreshDefaultModelProviderList();
      });

      // Assert that setModelProviderConfig was not called
      const azure = result.current.defaultModelProviderList.find((m) => m.id === 'azure');
      expect(azure?.chatModels).toEqual([{ id: 'abc', deploymentName: 'abc' }]);
    });

    it('openai', () => {
      const { result } = renderHook(() => useUserStore());
      act(() => {
        useUserStore.setState({
          serverLanguageModel: {
            openai: {
              enabled: true,
              enabledModels: ['gpt-4-0125-preview', 'gpt-4-turbo-2024-04-09'],
              serverModelCards: [
                {
                  displayName: 'ChatGPT-4',
                  functionCall: true,
                  id: 'gpt-4-0125-preview',
                  contextWindowTokens: 128000,
                  enabled: true,
                },
                {
                  displayName: 'ChatGPT-4 Vision',
                  functionCall: true,
                  id: 'gpt-4-turbo-2024-04-09',
                  contextWindowTokens: 128000,
                  vision: true,
                  enabled: true,
                },
              ],
            },
          },
        });
      });

      act(() => {
        result.current.refreshDefaultModelProviderList();
      });

      // Assert that setModelProviderConfig was not called
      const openai = result.current.defaultModelProviderList.find((m) => m.id === 'openai');
      expect(openai?.chatModels).toEqual([
        {
          displayName: 'ChatGPT-4',
          enabled: true,
          functionCall: true,
          id: 'gpt-4-0125-preview',
          contextWindowTokens: 128000,
        },
        {
          displayName: 'ChatGPT-4 Vision',
          enabled: true,
          functionCall: true,
          id: 'gpt-4-turbo-2024-04-09',
          contextWindowTokens: 128000,
          vision: true,
        },
      ]);
    });
  });

  describe('refreshModelProviderList', () => {
    it('visible', async () => {
      const { result } = renderHook(() => useUserStore());
      act(() => {
        useUserStore.setState({
          settings: {
            languageModel: {
              ollama: { enabledModels: ['llava'] },
            },
          },
        });
      });

      act(() => {
        result.current.refreshModelProviderList();
      });

      const ollamaList = result.current.modelProviderList.find((r) => r.id === 'ollama');
      // Assert that setModelProviderConfig was not called
      const model = ollamaList?.chatModels.find((c) => c.id === 'llava');

      expect(model).toMatchSnapshot();
    });

    it('modelProviderListForModelSelect should return only enabled providers', () => {
      const { result } = renderHook(() => useUserStore());

      act(() => {
        useUserStore.setState({
          settings: {
            languageModel: {
              perplexity: { enabled: true },
              azure: { enabled: false },
            },
          },
        });
      });

      act(() => {
        result.current.refreshModelProviderList();
      });

      const enabledProviders = modelProviderSelectors.modelProviderListForModelSelect(
        result.current,
      );
      expect(enabledProviders).toHaveLength(3);
      expect(enabledProviders.at(-1)!.id).toBe('perplexity');
    });
  });

  describe('removeEnabledModels', () => {
    it('should remove the specified model from enabledModels', async () => {
      const { result } = renderHook(() => useUserStore());
      const model = 'gpt-3.5-turbo';

      const spyOn = vi.spyOn(userService, 'updateUserSettings');

      act(() => {
        useUserStore.setState({
          settings: {
            languageModel: {
              azure: { enabledModels: ['gpt-3.5-turbo', 'gpt-4'] },
            },
          },
        });
      });

      await act(async () => {
        console.log(JSON.stringify(result.current.settings));
        await result.current.removeEnabledModels('azure', model);
      });

      expect(spyOn).toHaveBeenCalledWith(
        { languageModel: { azure: { enabledModels: ['gpt-4'] } } },
        expect.any(AbortSignal),
      );
    });
  });

  describe('toggleEditingCustomModelCard', () => {
    it('should update editingCustomCardModel when params are provided', () => {
      const { result } = renderHook(() => useUserStore());

      act(() => {
        result.current.toggleEditingCustomModelCard({ id: 'test-id', provider: 'openai' });
      });

      expect(result.current.editingCustomCardModel).toEqual({ id: 'test-id', provider: 'openai' });
    });

    it('should reset editingCustomCardModel when no params are provided', () => {
      const { result } = renderHook(() => useUserStore());

      act(() => {
        result.current.toggleEditingCustomModelCard();
      });

      expect(result.current.editingCustomCardModel).toBeUndefined();
    });
  });

  describe('toggleProviderEnabled', () => {
    it('should enable the provider', async () => {
      const { result } = renderHook(() => useUserStore());

      await act(async () => {
        await result.current.toggleProviderEnabled('minimax', true);
      });

      expect(userService.updateUserSettings).toHaveBeenCalledWith(
        { languageModel: { minimax: { enabled: true } } },
        expect.any(AbortSignal),
      );
    });

    it('should disable the provider', async () => {
      const { result } = renderHook(() => useUserStore());
      const provider = 'openai';

      await act(async () => {
        await result.current.toggleProviderEnabled(provider, false);
      });

      expect(userService.updateUserSettings).toHaveBeenCalledWith(
        { languageModel: { openai: { enabled: false } } },
        expect.any(AbortSignal),
      );
    });
  });

  describe('updateEnabledModels', () => {
    // TODO: 有待 updateEnabledModels 实现的同步改造
    it('should add new custom model to customModelCards', async () => {
      const { result } = renderHook(() => useUserStore());
      const provider = 'openai';
      const modelKeys = ['gpt-3.5-turbo', 'custom-model'];
      const options = [{ value: 'gpt-3.5-turbo' }, {}];

      await act(async () => {
        await result.current.updateEnabledModels(provider, modelKeys, options);
      });

      expect(userService.updateUserSettings).toHaveBeenCalledWith(
        {
          languageModel: {
            openai: {
              customModelCards: [{ id: 'custom-model' }],
              // TODO：目标单测中需要包含下面这一行
              // enabledModels: ['gpt-3.5-turbo', 'custom-model'],
            },
          },
        },
        expect.any(AbortSignal),
      );
    });

    it('should not add removed model to customModelCards', async () => {
      const { result } = renderHook(() => useUserStore());
      const provider = 'openai';
      const modelKeys = ['gpt-3.5-turbo'];
      const options = [{ value: 'gpt-3.5-turbo' }];

      act(() => {
        useUserStore.setState({
          settings: {
            languageModel: {
              openai: { enabledModels: ['gpt-3.5-turbo', 'gpt-4'] },
            },
          },
        });
      });

      await act(async () => {
        await result.current.updateEnabledModels(provider, modelKeys, options);
      });

      expect(userService.updateUserSettings).toHaveBeenCalledWith(
        {
          languageModel: { openai: { enabledModels: ['gpt-3.5-turbo'] } },
        },
        expect.any(AbortSignal),
      );
    });
  });

  describe('useFetchProviderModelList', () => {
    it('should fetch data when enabledAutoFetch is true', async () => {
      const { result } = renderHook(() => useUserStore());
      const provider = 'openai';
      const enabledAutoFetch = true;

      const spyOn = vi.spyOn(result.current, 'refreshDefaultModelProviderList');

      vi.spyOn(modelsService, 'getModels').mockResolvedValueOnce([]);

      renderHook(() => result.current.useFetchProviderModelList(provider, enabledAutoFetch));

      await waitFor(() => {
        expect(spyOn).toHaveBeenCalled();
      });

      // expect(result.current.settings.languageModel.openai?.latestFetchTime).toBeDefined();
      // expect(result.current.settings.languageModel.openai?.remoteModelCards).toBeDefined();
    });

    it('should not fetch data when enabledAutoFetch is false', async () => {
      const { result } = renderHook(() => useUserStore());
      const provider = 'openai';
      const enabledAutoFetch = false;

      const spyOn = vi.spyOn(result.current, 'refreshDefaultModelProviderList');

      vi.spyOn(modelsService, 'getModels').mockResolvedValueOnce([]);

      renderHook(() => result.current.useFetchProviderModelList(provider, enabledAutoFetch));

      await waitFor(() => {
        expect(spyOn).not.toHaveBeenCalled();
      });
    });
  });
});
