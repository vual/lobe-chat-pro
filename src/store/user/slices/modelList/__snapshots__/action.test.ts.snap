// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LLMSettingsSliceAction > refreshModelProviderList > visible 1`] = `
{
  "contextWindowTokens": 4096,
  "description": "LLaVA 是结合视觉编码器和 Vicuna 的多模态模型，用于强大的视觉和语言理解。",
  "displayName": "LLaVA 7B",
  "enabled": true,
  "id": "llava",
  "vision": true,
}
`;
